sw/host/opentitanlib/src/transport/ti50emulator/gpio.rs

```diff
@@ -241,148 +204,86 @@ impl Ti50GpioPin {
     fn update_dut_state(&self) -> Result<()> {
         if let Some(ref mut pin_fd) = *self.socket.borrow_mut() {
             let mut buf: [u8; GPIO_BUF_SIZE] = [0; GPIO_BUF_SIZE];
-            pin_fd.write_all(&[GpioStreamCommand::GetState as u8])?;
+            pin_fd.write_all(&[b'?'])?;
             let len = pin_fd.read(&mut buf[..])?;
-            let mut state = self.dut_state.get();
-            state.update(&buf[0..len])?;
-            self.dut_state.set(state);
+            if len > 0 {
+                self.dut_state.set(Logic::try_from(buf[0])?);
+            }
```

Comment:
I do not understand you comment completely.  But you are right that there is a problem here.

In the new protocol, the emulator will send one of these file characters: '0', '1', 'H', 'L', 'Z'. And a single character will completely describe how the emulator is pulling "its end" of the wire.  When one character is received, that completely supersedes anything received before.

Thus, we should pick that _last_ character not the first, if multiple characters happen to be delivered through one call to `read()`.

Reply:
Maybe I didn't make myself clear, you should decode all bytes in loop. Missing bytes will result in non-deterministic tests result. Checking the last byte is just a compromise. For slow changes in the state of the wire there is a chance that this implementation will work correctly, but in the case of a fast changing signal where there is a greater chance of sending many bytes to the pipe it will fail.

Reply:
I still have trouble understanding.  Are you saying that I need to loop over the bytes in `buf`, rather than merely looking at the last one.  I do not understand what `update_dut_state()` would do with these pieces of information, other than the last, most recent, one.

If you are instead saying that I need to make multiple calls to `read()` in order to ensure that we have received the last status.  That is a valid point.  However, it is hard to know for how long to wait for the reply, since there is no way to tell the difference between replies and unsolicited status updates.  Putting the socket in non-blocking mode, and reading until it returns WouldBlock is not a guarantee.
Rather than enhancing the protocol to tie requests and response by some sequence number, I would rather follow this CL up with one that does away with the query request/response altogether, and instead relies only on the unsolicited status updates from the sub-process.  I imagine a protocol in which upon establishing a socket, the sub-process sends a single character with its current state, and from then on sends updates as they happen.  This parent process will use `mio` to listen to all the sockets all the time, even when there is no calls made to the `Gpio` objects.  This would allow implementation of `GpioMonitoring`, which our tests rely on.

Reply:
As i said before checking last byte is compromise.
What i am afraid is test instability. 
(My doubts are based on the assumption of non-deterministic process scheduling.)

For example: 
If OT host App wait for state '1' and buffer contains 2 bytes [10] checking only last give false result.
If TockOS App is fast enough OS can accumulate more that 1 byte in the stream. Sometimes you will get [1] and second time  [10] with make test unstable. 

It can be avoided by synchronizing Host and Device app or providing some information about time.
I don't know how test will look like, maybe we should check Rising/Falling edge or wire level within time range.

Reply:
Ok, think I understand.

I do not agree that your example of host app (test script) waiting for "1" and buffer containing "10" can be reasonably "solved" by looking at all elements of the array.  What do you propose we should return, that the level of the GPIO is high? even though we know that the code under test has already taken it low again.

My goal is to be able to support mostly the same kinds of automatic test with host emulation, as we can support on the physical development board.  If we are looking for a very narrow pulse on a GPIO pin, then the `Gpio` trait `read()` method is not adequate.  Repeated reading could miss such a pulse on a development board, and it is OK that it could also miss such a pulse on host emulation.  The test needs to use the `GpioMonitoring` trait to be able to reliably capture a transcript of GPIO level changes, to be sure to detect such a pulse.

It is true that with the current code, reading a GPIO is not guaranteed to get the current value, but could instead return some prior value of the GPIO, particularly if there has been several recent state changes. In other words if the test is expecting the current level to be '0', and the emulator has sent "10", then you are right that the current code could read just "1", and report high level, contrary to test expectations and actual emulator behavior.

In such a case, if reading the GPIO again, before long, we should get a low level.  This could be mitigated by a synchronization protocol.  I do not think that sequence number is required, actually, but if the emulator acknowledged each '?' query by first sending the character representing the current level, and then sent e.g. a '!', then this code would know to keep reading until it got a '!', and use the last Logic character before the '!' as the value to report.  This would be a minor change to the host emulator code.  Let me know what you think.